pkg/utils/gardener/project.go

```diff
@@ -46,15 +47,37 @@ func ProjectForNamespaceFromReader(ctx context.Context, reader client.Reader, na
 }
 
 // ProjectAndNamespaceFromReader returns the Project responsible for a given <namespace>. It reads the namespace and
-// fetches the project name label. Then it will read the project with the respective name.
+// fetches the project name label. Then it will read the project with the respective name or returns error if there is no
+// project for the given namespace.
 func ProjectAndNamespaceFromReader(ctx context.Context, reader client.Reader, namespaceName string) (*gardencorev1beta1.Project, *corev1.Namespace, error) {
 	namespace := &corev1.Namespace{}
 	if err := reader.Get(ctx, kutil.Key(namespaceName), namespace); err != nil {
 		return nil, nil, err
 	}
 
-	projectName := namespace.Labels[v1beta1constants.ProjectName]
-	if projectName == "" {
+	projectName, ok := namespace.Labels[v1beta1constants.ProjectName]
+	if !ok {
+		return nil, namespace, fmt.Errorf("project namespace doesn't have required label")
+	}
+
+	project := &gardencorev1beta1.Project{}
+	if err := reader.Get(ctx, kutil.Key(projectName), project); err != nil {
+		return nil, namespace, err
+	}
+
+	return project, namespace, nil
+}
+
+// NamespaceBelongsToAnyProject returns the Project responsible for a given <namespace>. It reads the namespace and
+// fetches the project name label. Then it will read the project with the respective name.
+func NamespaceBelongsToAnyProject(ctx context.Context, reader client.Reader, namespaceName string) (*gardencorev1beta1.Project, *corev1.Namespace, error) {
```

Comment:
Instead of duplicating the function, can't we handle this corner case in the invocations of the `gutil.ProjectAndNamespaceFromReader` func? For example

```diff
diff --git a/pkg/operation/garden/garden.go b/pkg/operation/garden/garden.go
index 619b2a274..043881c85 100644
--- a/pkg/operation/garden/garden.go
+++ b/pkg/operation/garden/garden.go
@@ -55,7 +55,14 @@ func (b *Builder) WithProject(project *gardencorev1beta1.Project) *Builder {
 func (b *Builder) WithProjectFrom(reader client.Reader, namespace string) *Builder {
        b.projectFunc = func(ctx context.Context) (*gardencorev1beta1.Project, error) {
                project, _, err := gutil.ProjectAndNamespaceFromReader(ctx, reader, namespace)
-               return project, err
+               if err != nil {
+                       return nil, err
+               }
+               if project == nil {
+                       return nil, fmt.Errorf("cannot find Project for namespace '%s'", namespace)
+               }
+
+               return project, nil
        }
        return b
 }
```

It can be the case some other invocation of `ProjectAndNamespaceFromReader` that needs to be adapted - the one in shoot_control.go of gardenlet.